torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
   }
 }
 
+bool shouldConsiderForMerge(detail::Vertex<Node*>* v) {
+  if (v->rdata.size() >= 2) {
+    return true;
+  }
+  JIT_ASSERT(v->rdata.size() == 1);
+  auto * node = *v->rdata.begin();
+  if (node->kind() == prim::Constant) {
+    return false;
+  }
+  return isDifferentiable(node);
+}
+
+static detail::DynamicDAG<Node*> make_dependency_graph(Block * block) {
+  detail::DynamicDAG<Node*> dag;
+  std::unordered_map<Node*,detail::Vertex<Node*>*> node_to_vertex;
+  // NB: the block's param and return nodes are not in the dependency graph.
+  for (Node * node : block->nodes()) {
+    node_to_vertex[node] = dag.newVertex(node);
+  }
+  for (auto * node : block->nodes()) {
+    for (auto * v : node->outputs()) {
+      for (auto & use : v->uses()) {
+        // [Determine data dependencies]
+        // Consider the following code:
+        //     y = f(x)
+        //     if k:
+        //        w += y
+        //     z = g(y)
+        // This produces a dependency graph with 3 vertices:
+        // (0: f)   (1: if k ...)   (2: g)
+        // We need to peek into the if Node* to determine its data dependencies
+        // (the body depends on the output of f, so Vertex 1 depends on Vertex 0).
+        // For each Use of y, we find an owning node of y that is a part of the
+        // dependency graph (in this case, the Vertex containing the if Node*)
+        // and then record the dependency.
+        auto * owning_node = use.user;
+        while (owning_node != nullptr) {
```

Comment:
Shouldn't reaching `nullptr` be an error here? We should eventually end up in the current scope. Can you assert on it not happening after the loop? (to make sure we exit via break)

Reply:
I wasn't sure about how param nodes and return nodes work. Currently the param node and return node of a graph aren't in DynamicDAG (maybe they should be). 

If they aren't in DynamicDAG and if the use.user is the return node, then owning_node can go up to nullptr. I agree that this is weird and I'll try to see if I can figure out what is going on here

Reply:
True, that sounds reasonable